[minor] AI broker ansible playbook integration

karol-czarnecki opened the pull request:
https://jsw.ibm.com/browse/MASHP-1612


This PR provide support to deploy AI Broker and components to cluster with MAS Core 9.x installed. 

There are 3 new roles: 

1. Role: `odh`
```
- Install Red Hat OpenShift Serverless Operator
- Install Red Hat OpenShift Service Mesh Operator
- Install Authorino Operator
- Install Open Data Hub Operator
- Create DSCInitialization instance
- Create Data Science Cluster
- Create Create Data Science Pipelines Application
```
2. Role: `kmodels`
```
- Install Kmodel controller
- Install istio
- Install Kmodel store
- Install Kmodel watcher
```
3. Role: `aibroker`
```
- Install AI Broker api application
- Create AI Broker tenant
- Create, delete AI Broker API Key
- Create, delete AWS S3 API Key
- Create, delete WatsonX AI API Key
```
4. Screenshots
- Installed operators
![image](https://github.com/ibm-mas/ansible-devops/assets/112193189/41e28425-9d8e-4821-bf2d-1ffabfef32a7)
- pods 
![image](https://github.com/ibm-mas/ansible-devops/assets/112193189/58d967bf-84b6-43ff-9cdd-b4205dd92fee)
- routes 
![image](https://github.com/ibm-mas/ansible-devops/assets/112193189/43f539f2-2c63-43c8-b001-bfd94ec9f489)
- cr 
![image](https://github.com/ibm-mas/ansible-devops/assets/112193189/3bd403d2-9201-4936-9fee-ab03005059f6)

sanjayprab commented on docs/playbooks/oneclick-aibroker.md:
```diff
@@ -0,0 +1,216 @@
+
+## Prerequisites
+
+You will need a RedHat OpenShift v4.12 cluster with IBM Maximo Application Suite Core v9.x already be installed, the [oneclick-core](oneclick-core.md) playbook can be used to set this up.
```
this doesn't work on 4.13 and 4.14

karol-czarnecki replied:
@sanjayprab this one is resolved

whitfiea reviewed (changes requested):
@karol-czarnecki The PR needs some more words describing what this change is and how it will work, plus something that shows that this has been tested etc.

Also the new roles have no readme in the contents and are not included in the docs.

whitfiea commented on ibm/mas_devops/roles/suite_app_install/vars/aibroker.yml:
```diff
@@ -0,0 +1,7 @@
+---
+# Application specific variables for aibroker
+mas_app_fqn: aibrokerapps.apps.mas.ibm.com
```
AIBroker isn't a MAS app so I don't think this should be included

karol-czarnecki replied:
this one is resolved

whitfiea commented on ibm/mas_devops/roles/aibroker/files/create_s3.sh:
Why do we have these files (and other sh scripts) here when the changes could be done in ansible?

karol-czarnecki replied:
We are aware that we use shell scripts here - we are going to create work items to move this to ansible in future

whitfiea commented on ibm/mas_devops/roles/odh/templates/aibroker/storage-class.yml.j2:
Don't you need an NFS Server to be able to use this storageclass?

karol-czarnecki replied:
this is removed as we don't use it anymore

whitfiea commented on ibm/mas_devops/playbooks/oneclick_add_aibroker.yml:
```diff
@@ -0,0 +1,29 @@
+          - lookup('env', 'IBM_ENTITLEMENT_KEY') != ""
+          - lookup('env', 'MAS_INSTANCE_ID') != ""
+          - lookup('env', 'APP_DOMAIN') != ""
+          - lookup('env', 'ARTIFACTORY_USERNAME') != ""
```
ARTIFACTORY is internal only and shouldn't be a requirement.

whitfiea commented on ibm/mas_devops/roles/kmodels/templates/kmodels-store/deployment.yml.j2:
```diff
@@ -0,0 +1,43 @@
+      imagePullSecrets:
+        # - name: regcred
+        - name: {{ pullSecretName }}
+      containers:
```
is there is no resource request/limits or an liveness/readiness probes>?

karol-czarnecki replied:
this is fixed

whitfiea commented on ibm/mas_devops/roles/kmodels/templates/routes/controller.yml.j2:
```diff
@@ -0,0 +1,16 @@
+    weight: 100
+  port:
+    targetPort: http
+  tls:
```
this needs to re-encrypt the connection to the backend

karol-czarnecki replied:
We are waiting for research team to confirm if there is no issues for set security context for watcher - we will create work item to fix this in future

karol-czarnecki replied:
Hi @whitfiea we don't create routes right now

karol-czarnecki commented:
> @karol-czarnecki The PR needs some more words describing what this change is and how it will work, plus something that shows that this has been tested etc.
> 
> Also the new roles have no readme in the contents and are not included in the docs.

Hi @whitfiea 
Added more description to this PR, also added docs for 3 new roles and updated readme for playbook

karol-czarnecki commented on ibm/mas_devops/roles/kmodels/templates/kmodels-controller/rbac.yml.j2:
```diff
@@ -0,0 +1,37 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: km-controller-role
```
Hi @whitfiea 
For this issue - we are waiting for research team to test because we don't know if we change this to narrow role our code will work. 

We recall that km-controller needs to have access to other namespace. 

We are going to create work item for this and fix this in future

karol-czarnecki replied:
Hi @whitfiea ClusterRole removed

karol-czarnecki commented on ibm/mas_devops/roles/kmodels/templates/kmodels-store/network.yml.j2:
```diff
@@ -0,0 +1,14 @@
+  namespace: mas-{{ mas_instance_id }}-aibroker
+spec:
+  ports:
+    - name: http
```
Hi @whitfiea 
There is no sensitive data in store so there is no need to use secure port - we will create work item to fix this in future

whitfiea replied:
@karol-czarnecki so there is no authentication that occurs and so anyone can call the route for any of these apis? That doesn't seem right to me and all connections should be secure in one way or another.

durera replied:
The operator should be creating these services anyway, not the ansible code, whose job should just be to install the operator and create the operand.  The whole thing needs a rewrite frankly because it doesn't meet the MAS application contract (if it did we wouldn't have this special install role and would be able to re-use the standard mas_app_install, mas_app_cfg roles), but we're shipping it as-is anyway.

karol-czarnecki commented on ibm/mas_devops/roles/kmodels/templates/kmodels-watcher/deployment.yml.j2:
```diff
@@ -0,0 +1,40 @@
+      imagePullSecrets:
+        # - name: regcred
+        - name: {{ pullSecretName }}
+      containers:
```
We are waiting for research team to confirm if there is no issues for set security context for watcher - we will create work item to fix this in future

karol-czarnecki replied:
request/limits or probes are done - we would like to move securityContext fix for future

karol-czarnecki commented on ibm/mas_devops/roles/kmodels/templates/routes/store.yml.j2:
```diff
@@ -0,0 +1,16 @@
+    name: km-store
+    weight: 100
+  port:
+    targetPort: http
```
There is no sensitive data in store so there is no need to use secure port - we will create work item to fix this in future

whitfiea replied:
How do calls authenticate with this API if there is no sensitive data? What is in this store?

karol-czarnecki replied:
Hi @whitfiea 
The store includes the model templates definitions. A template is a set of instructions how to create a model. It doesn't include any customer or sensitive data.

karol-czarnecki replied:
routes removed

karol-czarnecki commented on ibm/mas_devops/roles/kmodels/vars/main.yml:
```diff
@@ -0,0 +1,16 @@
+registry: "{{ lookup('env','ICR_CP') | default('docker-na-public.artifactory.swg-devops.com/wiotp-docker-local', true) }}"
+pullSecretName: "ibm-entitlement"
+# controller
+controllerTag: 1.0.0-pre.dev
```
this is fixed

karol-czarnecki commented on ibm/mas_devops/roles/odh/meta/main.yml:
```diff
@@ -0,0 +1,22 @@
+    - ibm
+    - mas
+    - devops
+    - mongodb
```
removed mongodb

whitfiea replied:
The file as a whole still has references to mongodb

karol-czarnecki replied:
this is done is 3 roles

karol-czarnecki commented on ibm/mas_devops/roles/odh/tasks/determine-storage-classes.yml:
```diff
@@ -0,0 +1,32 @@
+
+- name: "determine-storage-classes : Assert that primary storage class has been defined"
+  assert:
+    that: cpd_primary_storage_class is defined and cpd_primary_storage_class != ""
```
we are using this for define storage classes because then we use this specific class in our deployment for example: https://github.com/ibm-mas/ansible-devops/blob/04e507bd313ab7f35cbb74e5eb3bc7112217ff63/ibm/mas_devops/roles/odh/templates/odh/dspa-aws.yml.j2#L37

karol-czarnecki replied:
this is fixed now

whitfiea commented on ibm/mas_devops/roles/kmodels/vars/main.yml:
```diff
@@ -0,0 +1,27 @@
+storeTag: 1.0.0-pre.10xdev
+store: "aibroker/store:{{ storeTag }}"
+# watcher
+watcherTag: 1.0.0-pre.10xdev
```
We would need to remove all the pre releases in this file

karol-czarnecki replied:
changed to release versions

whitfiea commented on ibm/mas_devops/roles/aibroker/templates/tenant/connector-configmap.yml.j2:
```diff
@@ -0,0 +1,9 @@
+  namespace: mas-{{ mas_instance_id }}-aibroker
+data:
+  _example: >
+    Set here common configuration to all connectors in tenant namespace.
```
is this expected? that we have _exmaple?

whitfiea commented on docs/playbooks/oneclick-aibroker.md:
```diff
@@ -0,0 +1,267 @@
+
+## Overview
+
+This playbook will add **AI Broker v1.0.x** to an existing IBM Maximo Application Suite Core installation or as standalone instance.
```
This says "as a standalone instance" but the Prerequisites above say it should have MAS Core already installed?

karol-czarnecki replied:
it is fixed now

durera commented on ibm/mas_devops/roles/suite_app_install/vars/aibroker.yml:
What purpose does this var file (and the default spec) play?  Isn't the `aibroker` role what is doing the install?

karol-czarnecki replied:
Hi @durera 
Files removed

durera reviewed (approved):
Merging as-is, aibroker doesn't meet the requirements of a MAS application and is a special case hence the standalone role for it's install.  Hopefully this will be addressed in the near-future.
